File: method/process/cpu/kskipcg.py
import sys

import numpy as np
from numpy import dot
from numpy.linalg import norm
from mpi4py import MPI

if __name__ == "__main__":
    sys.path.append('../../../../')
    from krylov.method.common import getConditionParams
    from krylov.method.process.cpu.common import init, start, end, matvec, vecvec 


def k_skip_cg(A, b, epsilon, k, T=np.float64):
    # 初期化
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    num_of_process = comm.Get_size()
    x, b_norm, N, max_iter, residual, num_of_solution_updates = init(A, b, T)
    # root 
    Ax = np.empty(N, T)  # matvec演算結果格納用
    Ar = np.zeros((k + 2, N), T)
    Ap = np.zeros((k + 3, N), T)
    a = np.zeros(2 * k + 2, T)
    f = np.zeros(2 * k + 4, T)
    c = np.zeros(2 * k + 2, T)
    # local
    local_N = N // num_of_process
    local_A = np.empty((local_N, N), T)
    local_Ax = np.empty(local_N, T)  # matvec演算結果格納用
    local_Ar = np.zeros((k + 2, local_N), T)
    local_Ap = np.zeros((k + 3, local_N), T)
    local_a = np.zeros(2 * k + 2, T)
    local_f = np.zeros(2 * k + 4, T)
    local_c = np.zeros(2 * k + 2, T)

    # 初期残差
    Ar[0] = b - matvec(A, local_A, x, Ax, local_Ax, comm)
    Ap[0] = Ar[0]

    # 反復計算
    if rank == 0:
        start_time = start(method_name='k-skip CG', k=k)
    for i in range(max_iter):
        # 収束判定
        residual[i] = norm(Ar[0]) / b_norm
        isConverged = np.array([residual[i] < epsilon], dtype=bool)
        comm.Bcast(isConverged, root=0)
        if isConverged:
            break

        # 事前計算
        comm.Scatter(A, local_A, root=0)
        for j in range(1, k + 1):
            comm.Bcast(Ar[j-1], root=0)
            local_Ar = dot(local_A, Ar[j-1])
            comm.Gather(local_Ar, Ar[j], root=0)
        for j in range(1, k + 2):
            comm.Bcast(Ap[j-1], root=0)
            local_Ap = dot(local_A, Ap[j-1])
            comm.Gather(local_Ap, Ap[j], root=0)
        comm.Bcast(Ar)
        comm.Bcast(Ap)
        for j in range(2 * k + 1):
            jj = j // 2
            local_a[j] = dot(
                Ar[jj][rank * local_N: (rank+1) * local_N],
                Ar[jj + j % 2][rank * local_N: (rank+1) * local_N]
            )
        comm.Reduce(local_a, a, root=0)
        for j in range(2 * k + 4):
            jj = j // 2
            local_f[j] = dot(
                Ap[jj][rank * local_N: (rank+1) * local_N],
                Ap[jj + j % 2][rank * local_N: (rank+1) * local_N]
            )
        comm.Reduce(local_f, f, root=0)
        for j in range(2 * k + 2):
            jj = j // 2
            local_c[j] = dot(
                Ar[jj][rank * local_N: (rank+1) * local_N],
                Ap[jj + j % 2][rank * local_N: (rank+1) * local_N]
            )
        comm.Reduce(local_c, c, root=0)

        # CGでの1反復
        alpha = a[0] / f[1]
        beta = alpha ** 2 * f[2] / a[0] - 1
        x += alpha * Ap[0]
        Ar[0] -= alpha * Ap[1]
        Ap[0] = Ar[0] + beta * Ap[0]
        Ap[1] = matvec(A, local_A, Ap[0], Ax, local_Ax, comm)

        # CGでのk反復
        for j in range(k):
            for l in range(0, 2*(k-j)+1):
                a[l] += alpha*(alpha*f[l+2] - 2*c[l+1])
                d = c[l] - alpha*f[l+1]
                c[l] = a[l] + d*beta
                f[l] = c[l] + beta*(d + beta*f[l])

            # 解の更新
            alpha = a[0] / f[1]
            beta = alpha ** 2 * f[2] / a[0] - 1
            x += alpha * Ap[0]
            Ar[0] -= alpha * Ap[1]
            Ap[0] = Ar[0] + beta * Ap[0]
            Ap[1] = matvec(A, local_A, Ap[0], Ax, local_Ax, comm)

        num_of_solution_updates[i+1] = num_of_solution_updates[i] + k + 1

    else:
        isConverged = False

    num_of_iter = i + 1
    residual_index = i
    if rank == 0:
        end(start_time, isConverged, num_of_iter, residual, residual_index)


if __name__ == "__main__":
    A, b, epsilon, k, T = getConditionParams('condition.json')
    k_skip_cg(A, b, epsilon, k, T)

File: cpp/kskipcg.cpp
#include "MyBlas.h"

using namespace std;

template<typename T>
vector<T> kskipcg(const Matrix &A, const vector<T> &b, const double epsilon, const unsigned int k);

int main() {
  Matrix A = {
		{5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0}
	};
	vector<double> b = {3.0, 1.0, 4.0, 0.0, 5.0, -1.0, 6.0, -2.0, 7.0, -15.0};
  const double epsilon = 1e-8;
  const unsigned int k = 0; 

  auto x = kskipcg(A, b, epsilon, k);
  for(auto e: x) cout << e << endl;
}

template<typename T>
vector<T> kskipcg(const Matrix &A, const vector<T> &b, const double epsilon, const unsigned int k){
  const ul N = A.size();
  const ul max_iter = N*2;
  const double b_norm = vecvec(b, b);
	vector<T> x(N, 0.0);

  // 初期化
  vv Ar(k + 2, vector<double>(N, 0));
  vv Ap(k + 3, vector<double>(N, 0));
  vector<double> a(2 * k + 2, 0);
  vector<double> f(2 * k + 4, 0);
  vector<double> c(2 * k + 2, 0);

  // 初期残差
  Ar[0] = b - matvec(A, x);
  Ap[0] = Ar[0];

  // 反復計算
  ul i = 0;
  ul index = 0;
  while (i < max_iter) {
    // 収束判定
    double residual = vecvec(Ar[0], Ar[0]) / b_norm;
    if (residual < epsilon) break;

    // 事前計算
    ul jj;
    for (ui j=1;j<k+1;j++) Ar[j]=matvec(A, Ar[j-1]);
    for (ui j=1;j<k+2;j++) Ap[j]=matvec(A, Ap[j-1]);
    for (ui j=0;j<2*k+1;j++) { 
      jj = j;
      a[j] = vecvec(Ar[jj], Ar[jj + j % 2]);
    }
    for (ui j=0;j<2*k+4;j++) { 
      jj = j;
      f[j] = vecvec(Ap[jj], Ap[jj + j % 2]);
    }
    for (ui j=0;j<2*k+2;j++) { 
      jj = j;
      c[j] = vecvec(Ar[jj], Ap[jj + j % 2]);
    }

    // CGでの1反復
    double alpha = a[0] / f[1];
    double beta = pow(alpha,2) * f[2] / a[0] - 1;
    x = x + alpha * Ap[0];
    Ar[0] = Ar[0] - alpha * Ap[1];
    Ap[0] = Ar[0] + beta * Ap[0];
    Ap[1] = matvec(A, Ap[0]);

    // CGでのk反復
    for (ui j=0;j<k;j++) {
      for (ui l=0;l<2*(k-j)+1;l++) {
        a[l] += alpha*(alpha*f[l+2] - 2*c[l+1]);
        double d = c[l] - alpha*f[l+1];
        c[l] = a[l] + d*beta;
        f[l] = c[l] + beta*(d + beta*f[l]);
      }
      // 解の更新
      alpha = a[0] / f[1];
      beta = pow(alpha,2) * f[2] / a[0] - 1;
      x = x + alpha * Ap[0];
      Ar[0] = Ar[0] - alpha * Ap[1];
      Ap[0] = Ar[0] + beta * Ap[0];
      Ap[1] = matvec(A, Ap[0]);
    }

    i += (k + 1);
    index += 1;
  }

	return x;
}
